Moved schema loading and watch to new LocalSchemaLoader

jormaechea opened the pull request:
Resolves #27.

Decouples CLI layer from schema loading and watch feature.

Unit tests are still missing.

EdwardSalter commented on lib/schema-loaders/local-loader.js:
```diff
@@ -0,0 +1,55 @@
+
+	constructor(schemaPath) {
+		super();
+		this.schemaPath = schemaPath.match(/^\//) ? schemaPath : path.join(process.cwd(), schemaPath);
```
Correct me if I am wrong but the ternary statement here is checking to see if the path provided is an absolute path right? In which case it takes the path as is, otherwise it assumed it is relative and joins the path with the current working directory.

This regex will only work on *nix based systems and won't work on Windows.
Node has the `path.isAbsolute(dir)` [method](https://nodejs.org/api/path.html#path_path_isabsolute_path) available which should give a platform independent solution to that problem.

```suggestion
		this.schemaPath = path.isAbsolute(schemaPath) ? schemaPath : path.join(process.cwd(), schemaPath);
```

jormaechea replied:
Yeah, you're right. Looks like sometimes I forget that windows even exists.. 😂

EdwardSalter replied:
Ha. It's very easy to do.
It wasn't until I looked again that I realised that it was already like that in the CLI and the code was just ported over.
I guess it's not really been an issue since I am guessing that most people are running this under Docker or with relative paths.

EdwardSalter commented on lib/open-api-mocker.js:
```diff
@@ -5,53 +5,82 @@ const jsonRefs = require('json-refs');
 	setSchema(schema) {
-		this.schema = schema;
+		if(typeof schema === 'string') {
+			this.schemaLoader = new LocalSchemaLoader(schema);
```
Pulling out the schema loading functionality into a new class this does satisfy #27 since it allows both CLI and API usages to share this functionality 👍 .

I fail to see however how a custom schema loader is best implemented with these changes in place.
For instance, if we imagine a RemoteSchemaLoader that fetches a hosted schema via a network request and then periodically polls for changes (i.e. watch functionality).
The implementation of this could pretty much follow the exact interface that the LocalSchemaLoader method uses but there is no way to inject this class. A consumer would have to write this class, call it outside of OpenApiMocker and use the same `setSchema`, `validate` and `mock` method sequence that the CLI currently does outside of this PR.
Whilst this is not too much of an issue, a consumer would have to know this correct sequence to call if they needed to reload the schema based on certain conditions in their own handling of the schema.

I guess another option is to have a constructor parameter similar to the `server` option that allows a custom implementation to be injected or a fallback to the default implementation if not specified. Something like this:
```js
constructor({ schemaLoader }) {
  this.schemaLoader = schemaLoader || new LocalSchemaLoader();
}
```

jormaechea replied:
Yes, I didn't add the DI yet. I wanted to avoid generating conflicts with the custom Server implementation, so I was waiting for that to be ready. But definitely the constructor DI will be the way.

Regarding custom schema loaders, for example `RemoteSchemaLoader`, it must only implement a `load()` method that has to take care of schema loading (fetch) and watch feature (polling) in case it supports it.

I'll also write some docs to explain all this _extensions_ that can be made to the core features and how.

EdwardSalter reviewed (commented):
HI @jormaechea ,
I apologise that I have only just seen your request for a review today since I have not been at my PC over the Easter break.

I have reviewed the change and added a general comment as well as spotting something that I believe would be the cause of a bug whilst running under Windows.

jormaechea commented:
No problem. Just fixed the windows' bug in this PR and will keep moving forward with this, adding tests, docs and so on.

EdwardSalter commented on lib/open-api-mocker.js:
```diff
@@ -5,61 +5,90 @@ const jsonRefs = require('json-refs');
-	}
+		if(this.options.schemaLoader) {
+			const CustomSchemaLoader = this.options.schemaLoader;
+			this.schemaLoader = new CustomSchemaLoader(schema);
```
Instead of `options.schemaLoader` being a class, would it be possible to make it an already instantiated instance?
If an instance was passed to the constructor it would give the calling code the ability to control when it gets instantiated and how long the variable lives for.
A new class what want to be configurable (i.e. a PollingSchemaLoader may want a configurable amount of time between each poll) but because the calling code has no control over when the schema loader gets instantiated, there is no way to control those configurable properties. This was the reasoning behind the change to allow a custom Server object rather than a custom Server class.
For this to work I guess that the instantiated object would require a `setSchema` method that would be called here instead of the constructor.

EdwardSalter reviewed (commented):
I can see a lot of effort has gone into this pull request and the new separation of concerns is great.
I have left one comment regarding the new dependency injection regarding the custom schema loading functionality.

jormaechea commented:
@EdwardSalter I've been thinking about the best solution to the problem you presented, and I think I've found it (or a very good solution at least).

I introduced a new update, where you have to pass a SchemaLoader instance and changed their interface from this:

```ts
interface OpenApiSchemaLoader extends EventEmitter {
    constructor(schema: SchemaOption);
    load(): OpenApiSchema|Promise<OpenApiSchema>;
    watch(): void; // Optional
    unwatch(): void // Optional
}
```

to this

```ts
interface OpenApiSchemaLoader extends EventEmitter {
    load(schema: SchemaOption): OpenApiSchema|Promise<OpenApiSchema>;
    watch(): void; // Optional
}
```

As a cool side effect, SchemaLoaders are not destroyed and recreated on each schema change any more. This is a nice performance improvement.

I think we've reached a more-than-acceptable state here. I made some tests and everything is working as expected.
Do you wanna take another look at the code or try some of your real life cases with this branch? It would be really appreciated.

EdwardSalter reviewed (approved):
I think these newest changes are perfect and exactly how I envisioned it.

I have updated my project to use the latest version of this branch and it still works as expected (note: I am _not_ using a custom schema loader, only a custom server implementation).

jormaechea commented:
Great then, I'll merge it now
